homeassistant/components/openuv/coordinator.py

```diff
@@ -1,23 +1,94 @@
 from typing import Any, cast
 
-from pyopenuv.errors import OpenUvError
+from pyopenuv.errors import InvalidApiKeyError, OpenUvError
 
-from homeassistant.core import HomeAssistant
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant, callback
+from homeassistant.data_entry_flow import FlowResult
 from homeassistant.helpers.debounce import Debouncer
 from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
 
-from .const import LOGGER
+from .const import DOMAIN, LOGGER
 
 DEFAULT_DEBOUNCER_COOLDOWN_SECONDS = 15 * 60
 
 
+class InvalidApiKeyMonitor:
+    """Define a monitor for failed API calls (due to bad keys) across coordinators."""
+
+    DEFAULT_FAILED_API_CALL_THRESHOLD = 5
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize."""
+        self._count = 1
+        self._lock = asyncio.Lock()
+        self._reauth_flow_manager = ReauthFlowManager(hass, entry)
+        self.entry = entry
+
+    async def async_increment(self) -> None:
+        """Increment the counter."""
+        LOGGER.debug("Invalid API key response detected (number %s)", self._count)
+        async with self._lock:
+            self._count += 1
+            if self._count > self.DEFAULT_FAILED_API_CALL_THRESHOLD:
+                self._reauth_flow_manager.start_reauth()
+
+    async def async_reset(self) -> None:
+        """Reset the counter."""
+        async with self._lock:
+            self._count = 0
+            self._reauth_flow_manager.cancel_reauth()
+
+
+class ReauthFlowManager:
+    """Define an OpenUV reauth flow manager."""
+
+    def __init__(self, hass: HomeAssistant, entry: ConfigEntry) -> None:
+        """Initialize."""
+        self.entry = entry
+        self.hass = hass
+
+    @callback
+    def _get_active_reauth_flow(self) -> FlowResult | None:
+        """Get an active reauth flow (if it exists)."""
+        try:
+            [reauth_flow] = [
+                flow
+                for flow in self.hass.config_entries.flow.async_progress_by_handler(
+                    DOMAIN
+                )
+                if flow["context"]["source"] == "reauth"
+                and flow["context"]["entry_id"] == self.entry.entry_id
+            ]
+        except ValueError:
+            return None
+
+        return reauth_flow
+
+    @callback
+    def cancel_reauth(self) -> None:
+        """Cancel a reauth flow (if appropriate)."""
+        if reauth_flow := self._get_active_reauth_flow():
+            LOGGER.debug("API seems to have recovered; canceling reauth flow")
+            self.hass.config_entries.flow.async_abort(reauth_flow["flow_id"])
+
+    @callback
+    def start_reauth(self) -> None:
+        """Start a reauth flow (if appropriate)."""
+        if not self._get_active_reauth_flow():
+            LOGGER.debug("Multiple API failures in a row; starting reauth flow")
+            self.entry.async_start_reauth(self.hass)
```

Comment:
It really shouldn't be possible to get to a state where there is a reauth flow active but we can call reauth again.

Starting reauth should probably unload the config entry and put it into a `ConfigEntryState.SETUP_ERROR` state

Reply:
Won't unloading the entry prevent us from responding if the API recovers? From part 3:

> If the API recovers and returns an HTTP 200 response after the re-auth flow has been created, we assume that the prior HTTP 401/403 responses were a service issue and abort the re-auth flow.

Reply:
Short term I fixed the check that would allow multiple reauths for the same config entry

https://github.com/home-assistant/core/pull/81753

Reply:
I don't think we should unload the entry on reauth. Possibly re-load it so it's only every in it's setup stage. I think we should handle transient auth errors. If the error from an update coordinator goes away any started reauth should be cancelled.

Reply:
@elupus I'm unclear on:

> Just realize that is just what this integration is doing. Feels like that ought to be moved into main coordinator.

All of this occurs within a single object that our two coordinators share.

Reply:
> I don't think we should unload the entry on reauth.



> Won't unloading the entry prevent us from responding if the API recovers? From part 3:



I agree it doesn't make sense to unload it unless it happens during setup 👍

Reply:
> All of this occurs within a single object that our two coordinators share.

This makes sense to me since you'll never know which one is going to hit the auth problem first.